File: apps/web/src/lib/search/hybrid.ts
/**
 * Client-side hybrid search: fuses the pure-TS BM25 index with client-side
 * semantic similarity (transformers.js). This is the browser twin of
 * `services/search/app/hybrid.py` — same formula, same constants, two
 * independent implementations so the optional backend is never required.
 *
 * See docs/research/search-and-retrieval.md §4 for why a weighted linear
 * combination (rather than e.g. reciprocal rank fusion) was chosen: it
 * keeps the per-result score breakdown human-readable for Glass Mode.
 */
import { buildBm25Index, tokenize, type Bm25Index } from './bm25'
import { semanticScores } from './semantic'
import { loadEmbeddingsManifest, loadProducts } from './loadData'
import { BM25_B, BM25_K1, EMBEDDING_MODEL_NAME_OFFLINE, FUSION_ALPHA, LIMITATIONS, WHY_AI_WAS_USED } from './config'
import type { GlassExplanation, Product, SearchOutcome, SearchResultItem } from './types'

let indexPromise: Promise<{ index: Bm25Index; products: Product[] }> | null = null

function getIndex() {
  if (!indexPromise) {
    indexPromise = loadProducts().then((products) => {
      const docs = products.map((p) => tokenize(`${p.title} ${p.description}`))
      return { index: buildBm25Index(docs, BM25_K1, BM25_B), products }
    })
  }
  return indexPromise
}

function minMaxNormalize(values: ArrayLike<number>): Float64Array {
  let lo = Infinity
  let hi = -Infinity
  for (let i = 0; i < values.length; i++) {
    if (values[i] < lo) lo = values[i]
    if (values[i] > hi) hi = values[i]
  }
  const out = new Float64Array(values.length)
  const range = hi - lo
  for (let i = 0; i < values.length; i++) {
    out[i] = range < 1e-9 ? 0 : (values[i] - lo) / range
  }
  return out
}

export async function hybridSearch(query: string, limit = 12, alpha = FUSION_ALPHA): Promise<SearchOutcome> {
  const totalStart = performance.now()

  let t = performance.now()
  const queryTokens = tokenize(query)
  const tokenizeQueryMs = performance.now() - t

  t = performance.now()
  const { index, products } = await getIndex()
  const bm25Raw = index.score(queryTokens)
  const bm25ScoreMs = performance.now() - t

  t = performance.now()
  const { scores: semanticRaw } = await semanticScores(query)
  const embedAndScoreMs = performance.now() - t

  t = performance.now()
  const bm25Norm = minMaxNormalize(bm25Raw)
  const semanticNorm = minMaxNormalize(semanticRaw)
  const fused = new Float64Array(products.length)
  for (let i = 0; i < products.length; i++) {
    fused[i] = alpha * bm25Norm[i] + (1 - alpha) * semanticNorm[i]
  }
  const order = Array.from(fused.keys()).sort((a, b) => fused[b] - fused[a]).slice(0, limit)
  const fusionMs = performance.now() - t

  const results: SearchResultItem[] = order.map((idx, i) => ({
    product: products[idx],
    rank: i + 1,
    score: {
      bm25Raw: bm25Raw[idx],
      bm25Normalized: bm25Norm[idx],
      semanticRawCosine: semanticRaw[idx],
      semanticNormalized: semanticNorm[idx],
      fused: fused[idx],
    },
  }))

  const totalMs = performance.now() - totalStart
  const manifest = await loadEmbeddingsManifest()

  const glass: GlassExplanation = {
    provider: 'client',
    whyAiWasUsed: WHY_AI_WAS_USED,
    bm25K1: BM25_K1,
    bm25B: BM25_B,
    fusionAlpha: alpha,
    embeddingModel: EMBEDDING_MODEL_NAME_OFFLINE,
    embeddingDim: manifest?.dim ?? 384,
    embeddingSource: 'catalog vectors: precomputed offline artifact; query vector: embedded live in your browser via transformers.js (WebAssembly), never sent to a server',
    corpusSize: products.length,
    artifactGeneratedAt: manifest?.generated_at ?? null,
    timing: {
      tokenizeQueryMs,
      bm25ScoreMs,
      embedQueryMs: embedAndScoreMs,
      semanticScoreMs: 0,
      fusionMs,
      totalMs,
    },
    limitations: LIMITATIONS,
  }

  return { query, results, glass }
}

File: services/search/app/hybrid.py
"""Hybrid (BM25 + semantic) search over the product corpus.

Mirrors `apps/web/src/lib/search/hybrid.ts` so the server-side and
client-side implementations are two independent renderings of the same
documented algorithm (see docs/research/search-and-retrieval.md §4),
not two different search engines that happen to agree by coincidence.
"""

from __future__ import annotations

import time
from functools import lru_cache

import numpy as np
from rank_bm25 import BM25Okapi

from .config import BM25_B, BM25_K1, EMBEDDING_MODEL_NAME, FUSION_ALPHA
from .corpus import Corpus, load_corpus, tokenize
from .schemas import GlassExplanation, GlassTiming, ScoreBreakdown, SearchResultItem


@lru_cache(maxsize=1)
def _bm25_index() -> BM25Okapi:
    corpus = load_corpus()
    return BM25Okapi(corpus.tokenized_docs, k1=BM25_K1, b=BM25_B)


@lru_cache(maxsize=1)
def _embedding_model():
    from fastembed import TextEmbedding

    return TextEmbedding(model_name=EMBEDDING_MODEL_NAME)


def _min_max_normalize(values: np.ndarray) -> np.ndarray:
    lo, hi = float(values.min()), float(values.max())
    if hi - lo < 1e-9:
        return np.zeros_like(values)
    return (values - lo) / (hi - lo)


def search(query: str, limit: int, alpha: float = FUSION_ALPHA) -> tuple[list[SearchResultItem], GlassExplanation]:
    t_start = time.perf_counter()
    corpus: Corpus = load_corpus()

    t0 = time.perf_counter()
    query_tokens = tokenize(query)
    t_tokenize = time.perf_counter() - t0

    t0 = time.perf_counter()
    bm25_raw = np.asarray(_bm25_index().get_scores(query_tokens), dtype=np.float64)
    t_bm25 = time.perf_counter() - t0

    t0 = time.perf_counter()
    query_vector = next(_embedding_model().embed([query]))
    query_vector = np.asarray(query_vector, dtype=np.float32)
    t_embed = time.perf_counter() - t0

    t0 = time.perf_counter()
    # Embedding vectors are already L2-normalized, so the dot product is cosine similarity.
    semantic_raw = corpus.embedding_matrix @ query_vector
    t_semantic = time.perf_counter() - t0

    t0 = time.perf_counter()
    bm25_norm = _min_max_normalize(bm25_raw)
    semantic_norm = _min_max_normalize(semantic_raw)
    fused = alpha * bm25_norm + (1 - alpha) * semantic_norm
    order = np.argsort(-fused)[:limit]
    t_fusion = time.perf_counter() - t0

    results: list[SearchResultItem] = []
    for rank, idx in enumerate(order, start=1):
        product = corpus.products[idx]
        results.append(
            SearchResultItem(
                **{k: product[k] for k in ("id", "title", "description", "category", "brand",
                                            "price", "currency", "rating", "rating_count")},
                rank=rank,
                score=ScoreBreakdown(
                    bm25_raw=round(float(bm25_raw[idx]), 6),
                    bm25_normalized=round(float(bm25_norm[idx]), 6),
                    semantic_raw_cosine=round(float(semantic_raw[idx]), 6),
                    semantic_normalized=round(float(semantic_norm[idx]), 6),
                    fused=round(float(fused[idx]), 6),
                ),
            )
        )

    total = time.perf_counter() - t_start
    glass = GlassExplanation(
        bm25_k1=BM25_K1,
        bm25_b=BM25_B,
        fusion_alpha=alpha,
        embedding_model=EMBEDDING_MODEL_NAME,
        embedding_dim=int(corpus.embedding_matrix.shape[1]),
        embedding_source="precomputed offline artifact (models/search-embeddings/product_embeddings.json); query embedded live server-side",
        corpus_size=len(corpus.products),
        artifact_generated_at=corpus.embedding_manifest.get("generated_at"),
        timing=GlassTiming(
            tokenize_query_ms=round(t_tokenize * 1000, 3),
            bm25_score_ms=round(t_bm25 * 1000, 3),
            embed_query_ms=round(t_embed * 1000, 3),
            semantic_score_ms=round(t_semantic * 1000, 3),
            fusion_ms=round(t_fusion * 1000, 3),
            total_ms=round(total * 1000, 3),
        ),
    )
    return results, glass
